### ISSUE TITLE
0x73696d616f - The amount withdrawn by an user does not discount the fee paid which will leave funds stuck when calling `LidoVault::VaultEndedWithdraw()`

### ISSUE THREAD

**Issue by sherlock-admin4** [sherlock-audit/2024-08-saffron-finance-judging#87]
0x73696d616f

High

# The amount withdrawn by an user does not discount the fee paid which will leave funds stuck when calling `LidoVault::VaultEndedWithdraw()`

### Summary

In [LidoVault::vaultEndedWithdraw()](https://github.com/sherlock-audit/2024-08-saffron-finance/blob/main/lido-fiv/contracts/LidoVault.sol#L709), the pro-rata amount that the user has withdrawn is calculated as:
```solidity
variableToWithdrawnStakingEarningsInShares[msg.sender].mulDiv(
    vaultEndingETHBalance, vaultEndingStakesAmount
)
```
However, this does not take into account that the user paid fees, but `totalEarnings` does, which will lead to stuck ETH by overestimating the amount the user has withdrawn.

### Root Cause

In `LidoVault.sol:759`, `variableToWithdrawnStakingEarningsInShares[msg.sender]` does not discount the fees in shares the user has paid.

### Internal pre-conditions

None.

### External pre-conditions

None.

### Attack Path

1. Users withdraw via `LidoVault::vaultEndedWithdraw()`, but will receive less than they should and ETH will be stuck.

### Impact

ETH is stuck in the protocol and variable users take losses.

### PoC

Consider that 2 users have 50% of the variable bearer token each and the protocol starts with 100 ETH and 100 Shares.
After some time, the protocol accrues stEth and the holdings become 200 ETH and 100 shares.
Before the vault ends, user A withdraws his share of the rewards via `LidoVault::withdraw()`, which is:
```solidity
totalEarnings =  200 * (100 + 0) / 100 - 100 = 100
ethAmountOwed = totalEarnings / 2 = 100 / 2 = 50 // 50% of the bearer tokens, so divides by 2
protocolFee = 50 * 0.05 = 2.5
stakesAmountOwed = 50 / 2 = 25 // each share of stETH is worth 2 ETH, the same ratio as the protocol having 200 ETH and 100 shares.

// RESULT

totalProtocolFee = 2.5
variableToWithdrawnStakingEarningsInShares[userA] = 25
withdrawnStakingEarningsInStakes = 25
ETH = 150
shares = 75
```

Now, assume that the stETH shares double again its value, there are 300 ETH and 75 shares now in the protocol.
Consider that the vault ended so the fixed users claimed their 100 ETH, which leaves the protocol with 200 ETH and 50 shares (25 shares were removed to pay the fixed users).

And lastly, userB withdraws his variable rewards by calling `LidoVault::vaultEndedWithdraw()`. The vault has earnings that were not withdrawn, so in the beginning of `LidoVault::vaultEndedWithdraw()` it requests the entire balance of the contract and registers the following:
```solidity
vaultEndingStakesAmount = 50
vaultEndingETHBalance = 200
```
Then, `LidoVault::finalizeVaultEndedWithdrawals()` is called, which claims the withdrawals, getting 200 ETH and setting the following variables:
```solidity
amountWithdrawn = 200 ETH
fixedETHDeposit = 0 // was withdrawn by the fixed users already
vaultEndedStakingEarnings = 200 - 200 * 0.05 = 200 - 10 = 190
```
And finally, it calls again `LidoVault::vaultEndedWithdraw()` at the end, which leads to the following `totalEarnings` calculations:
```solidity
totalEarnings = 200 * 25 / 50 - 2.5 + 190 = 287.5
ethAmountOwed = totalEarnings / 2 = 287.5 / 2 = 143.75
```
Now, we calculate userA's amount to withdraw. We just need to subtract the shares of userA worth in ETH to the `totalEarnings / 2` to get his part:
```solidity
ethAmountOwed = 143.75 - 25 * 200 / 50 = 43.75
```

As userB has not withdraw, `variableToWithdrawnStakingEarningsInShares[userA] == 0`, so it gets `143.75`.

So summing up, they get `143.5 + 43.75 == 187.25`, which is less than 190 and some ETH is stuck.

### Mitigation

The already withdrawn amount should be subtract the shares of fees they already paid, that is:
```solidity
(variableToWithdrawnStakingEarningsInShares[msg.sender] - feeInShares[msg.sender]).mulDiv(
    vaultEndingETHBalance, vaultEndingStakesAmount
)
```

**Comment by sherlock-admin2** [comment]
The protocol team fixed this issue in the following PRs/commits:
https://github.com/saffron-finance/lido-fiv/pull/29
